lib/TPP/ConvertTppToLoops.cpp

```diff
@@ -22,49 +22,95 @@ using namespace mlir::tpp;
 
 namespace {
 
+static bool isScalarVal(Value val) { return !val.getType().isa<ShapedType>(); }
+
+static inline Value getZero(RewriterBase &rewriter, Location loc) {
+  return rewriter.create<arith::ConstantIndexOp>(loc, 0);
+}
+
+static inline Value getOne(RewriterBase &rewriter, Location loc) {
```

Comment:
Can you use `BuilderUtils.h::getConstIndex()` instead?

Reply:
Looks like build a constant with unknown location.

Reply:
It does, we don't need a location for compiler constants.

Reply:
No, it doesn't: https://github.com/plaidml/tpp-mlir/blob/1b7913f209b2154e876eef2f8c0855fb31cea6bf/lib/TPP/BuilderUtils.cpp#L23 We build constant op with unknown loc.

Reply:
I agreed with you: "It does (build a constant with unknown location)". :smile: 

We don't care about the location of compiler-generated constants.

Reply:
I am not completely sure why we don't care. We build a constant operation with no location. I did not see this pattern in MLIR. We can modify the method to accept a location too.

Reply:
> I am not completely sure why we don't care.

The main purpose of file locations is stepping through the code while debugging. 

When stepping through a line in the model (say, Python or C++), you can have multiple HW operations per IR instruction and multiple IR instructions per line of code. All operations need to be tagged with the same LOC as the program line, so that the debugger knows when to stop (breakpoint, single-step, etc).

Constants, however, are generally not operations. They're stored in the binary (some data section, constant pools), or converted to immediate values on instructions, or just optimised away. There's no way you can "set a breakpoint" or "step-through" a constant declaration, unless this becomes an actual operation (say, offload a CPU constant into GPU memory), which only really happens to tensors.

These particular constants are indices, integers and floats that will become part of instructions. A loop induction variable range check (usually immediate or zero check), a float zero-fill, which will be lowered as `XOR` or `MOV reg, zero`, etc.

None of them are relevant to the debugging experience, and therefore, need no location.

More generally, for actual ops, by blinding attaching "some" location to random constants and ops, we may be changing the semantics of what the debugger is expected to do.

A one-to-many rewrite can easily just use the original location to all the generated ops. A many-to-many rewrite needs to make sure what location is going where, as they may be in different (non-sequential) code lines. 

Ultimately, an op without location won't break the debugger more than a wrong location op. But none of that applies to compiler-generated constants.

> I did not see this pattern in MLIR.

Generic MLIR passes cannot assume where their constants came from. We can.

> We can modify the method to accept a location too.

If you do this, then make sure to default to `UnknownLoc`, so that it doesn't change the existing uses.